video/qtvr_decoder.cpp

```diff
@@ -593,6 +731,145 @@ void QuickTimeDecoder::PanoTrackHandler::initPanorama() {
+				int32 p11 = sourceSurface->getPixel(x1, y1);
+
+				uint8 a00, r00, g00, b00;
+				uint8 a01, r01, g01, b01;
+				uint8 a10, r10, g10, b10;
+				uint8 a11, r11, g11, b11;
+
+				sourceSurface->format.colorToARGB(p00, a00, r00, g00, b00);
+				sourceSurface->format.colorToARGB(p01, a01, r01, g01, b01);
+				sourceSurface->format.colorToARGB(p10, a10, r10, g10, b10);
+				sourceSurface->format.colorToARGB(p11, a11, r11, g11, b11);
+
+				int32 avgPixel00 = p00;
+				int32 avgPixel01 = sourceSurface->format.ARGBToColor((a00 + a01) >> 1, (r00 + r01) >> 1, (g00 + g01) >> 1, (b00 + b01) >> 1);
+				int32 avgPixel10 = sourceSurface->format.ARGBToColor((a00 + a10) >> 1, (r00 + r10) >> 1, (g00 + g10) >> 1, (b00 + b10) >> 1);
+				int32 avgPixel11 =  sourceSurface->format.ARGBToColor(
+					(a00 + a01 + a10 + a11) >> 2,
+					(r00 + r01 + r10 + r11) >> 2,
+					(g00 + g01 + g10 + g11) >> 2,
+					(b00 + b01 + b10 + b11) >> 2);
+
+				target->setPixel(x * 2, y * 2, avgPixel00);
+				target->setPixel(x * 2, y * 2 + 1, avgPixel01);
+				target->setPixel(x * 2 + 1, y * 2, avgPixel10);
+				target->setPixel(x * 2 + 1, y * 2 + 1,avgPixel11);
+			}
+		}
+	} else {
+		target->create(w * 2, h, sourceSurface->format);
+
+		for (uint y = 0; y < h; ++y) {
+			for (uint x = 0; x < w; ++x) {
+				uint x1 = (x + 1) % w;
+
+				int32 p00 = sourceSurface->getPixel(x, y);
+				int32 p01 = sourceSurface->getPixel(x1, y);
+
+				uint8 a00, r00, g00, b00;
+				uint8 a01, r01, g01, b01;
+				sourceSurface->format.colorToARGB(p00, a00, r00, g00, b00);
+				sourceSurface->format.colorToARGB(p01, a01, r01, g01, b01);
+
+				int32 avgPixel00 = p00;
+				int32 avgPixel01 = sourceSurface->format.ARGBToColor((a00 + a01) >> 1, (r00 + r01) >> 1, (g00 + g01) >> 1, (b00 + b01) >> 1);
+
+				target->setPixel(x * 2, y, avgPixel00);
+				target->setPixel(x * 2 + 1, y, avgPixel01);
+			}
+		}
+	}
+
+	return target;
+}
+
+void QuickTimeDecoder::PanoTrackHandler::boxAverage(Graphics::Surface *sourceSurface, int scaleFactor) {
+	uint16 h = sourceSurface->h;
+	uint16 w = sourceSurface->w;
+
+	// Apply anti-aliasing if quality is higher than 1
+	// Otherwise it'll be quicker to just copy the _planarProjection onto _projectedPano
+	if (scaleFactor == 1) {
+		_projectedPano->copyFrom(*sourceSurface);
+		return;
+	}
+
+	// Average out the pixels from the larger image
+	for (uint16 y = 0; y < h / scaleFactor; y++) {
+		for (uint16 x = 0; x < w / scaleFactor; x++) {
+			uint16 avgA = 0, avgR = 0, avgG = 0, avgB = 0;
+
+			for (uint8 row = 0; row < scaleFactor; row++) {
+				for (uint8 column = 0; column < scaleFactor; column++) {
+					uint8 a00, r00, g00, b00;
+					uint32 pixel00 = sourceSurface->getPixel(MIN((x * scaleFactor + row), w - 1), MIN((y * scaleFactor + column), h - 1));
+					_projectedPano->format.colorToARGB(pixel00, a00, r00, g00, b00);
+
+					avgA += a00;
+					avgR += r00;
+					avgG += g00;
+					avgB += b00;
+				}
+			}
+
+			uint8 scaleSquare = scaleFactor * scaleFactor;
```

Comment:
Please move it outside of all four loops, there is no need to compute it at every step, it will be constant.